File: TheReader/Models/HistoryItem.swift
import Foundation
import SwiftData

@Model
final class HistoryItem {
    @Attribute(.unique) var id: String
    var mangaId: String
    var mangaTitle: String
    var mangaCoverURL: String?
    var chapterId: String
    var chapterNumber: Double
    var chapterTitle: String
    var lastReadPage: Int
    var totalPages: Int
    var lastReadDate: Date

    var progress: Double {
        guard totalPages > 0 else { return 0 }
        return Double(lastReadPage) / Double(totalPages)
    }

    var progressPercentage: Int {
        Int(progress * 100)
    }

    var formattedChapter: String {
        if chapterNumber.truncatingRemainder(dividingBy: 1) == 0 {
            return "Chapter \(Int(chapterNumber))"
        }
        return "Chapter \(String(format: "%.1f", chapterNumber))"
    }

    var relativeDate: String {
        let calendar = Calendar.current
        if calendar.isDateInToday(lastReadDate) {
            return "Today"
        } else if calendar.isDateInYesterday(lastReadDate) {
            return "Yesterday"
        } else {
            let formatter = DateFormatter()
            formatter.dateFormat = "MMM d"
            return formatter.string(from: lastReadDate)
        }
    }

    init(
        id: String = UUID().uuidString,
        mangaId: String,
        mangaTitle: String,
        mangaCoverURL: String? = nil,
        chapterId: String,
        chapterNumber: Double,
        chapterTitle: String = "",
        lastReadPage: Int = 1,
        totalPages: Int = 1,
        lastReadDate: Date = .now
    ) {
        self.id = id
        self.mangaId = mangaId
        self.mangaTitle = mangaTitle
        self.mangaCoverURL = mangaCoverURL
        self.chapterId = chapterId
        self.chapterNumber = chapterNumber
        self.chapterTitle = chapterTitle
        self.lastReadPage = lastReadPage
        self.totalPages = totalPages
        self.lastReadDate = lastReadDate
    }
}

File: flutter_app/lib/models/history_item.dart
import 'package:hive/hive.dart';

part 'history_item.g.dart';

@HiveType(typeId: 4)
class HistoryItem extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String mangaId;

  @HiveField(2)
  final String mangaTitle;

  @HiveField(3)
  final String? mangaCoverUrl;

  @HiveField(4)
  final String chapterId;

  @HiveField(5)
  final double chapterNumber;

  @HiveField(6)
  final String chapterTitle;

  @HiveField(7)
  int lastReadPage;

  @HiveField(8)
  int totalPages;

  @HiveField(9)
  DateTime lastReadDate;

  HistoryItem({
    required this.id,
    required this.mangaId,
    required this.mangaTitle,
    this.mangaCoverUrl,
    required this.chapterId,
    required this.chapterNumber,
    this.chapterTitle = '',
    this.lastReadPage = 1,
    this.totalPages = 1,
    DateTime? lastReadDate,
  }) : lastReadDate = lastReadDate ?? DateTime.now();

  double get progress {
    if (totalPages == 0) return 0;
    return lastReadPage / totalPages;
  }

  int get progressPercentage => (progress * 100).toInt();

  String get formattedChapter {
    if (chapterNumber == chapterNumber.truncateToDouble()) {
      return 'Chapter ${chapterNumber.toInt()}';
    }
    return 'Chapter ${chapterNumber.toStringAsFixed(1)}';
  }

  String get relativeDate {
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);
    final readDate = DateTime(lastReadDate.year, lastReadDate.month, lastReadDate.day);

    if (readDate == today) {
      return 'Today';
    } else if (readDate == today.subtract(const Duration(days: 1))) {
      return 'Yesterday';
    } else {
      return '${lastReadDate.month}/${lastReadDate.day}';
    }
  }

  HistoryItem copyWith({
    String? id,
    String? mangaId,
    String? mangaTitle,
    String? mangaCoverUrl,
    String? chapterId,
    double? chapterNumber,
    String? chapterTitle,
    int? lastReadPage,
    int? totalPages,
    DateTime? lastReadDate,
  }) {
    return HistoryItem(
      id: id ?? this.id,
      mangaId: mangaId ?? this.mangaId,
      mangaTitle: mangaTitle ?? this.mangaTitle,
      mangaCoverUrl: mangaCoverUrl ?? this.mangaCoverUrl,
      chapterId: chapterId ?? this.chapterId,
      chapterNumber: chapterNumber ?? this.chapterNumber,
      chapterTitle: chapterTitle ?? this.chapterTitle,
      lastReadPage: lastReadPage ?? this.lastReadPage,
      totalPages: totalPages ?? this.totalPages,
      lastReadDate: lastReadDate ?? this.lastReadDate,
    );
  }
}
